src/stochastic_swap.rs

```diff
@@ -0,0 +1,257 @@
+            cost_reduced = false;
+            // Loop over edges of coupling graph
+            need_copy = true;
+            for idx in 0..num_edges {
+                let start_edge = edges[2 * idx];
+                let end_edge = edges[2 * idx + 1];
+                let start_qubit = trial_layout.phys_to_logic[start_edge];
+                let end_qubit = trial_layout.phys_to_logic[end_edge];
+                if qubit_set.contains(&start_qubit) && qubit_set.contains(&end_qubit) {
+                    // Try this edge to reduce cost
+                    if need_copy {
+                        new_layout = trial_layout.clone();
+                        need_copy = false;
+                    }
+                    new_layout.swap(start_edge, end_edge);
+                    // compute objective function
+                    new_cost = compute_cost(&scale.view(), &new_layout, gates, num_gates);
+                    // record progress if we succeed
+                    if new_cost < min_cost {
+                        cost_reduced = true;
+                        min_cost = new_cost;
+                        optimal_layout = new_layout.clone();
+                        optimal_start = start_edge;
+                        optimal_end = end_edge;
+                        optimal_start_qubit = start_qubit;
+                        optimal_end_qubit = end_qubit;
+                        need_copy = true;
+                    } else {
+                        new_layout.swap(start_edge, end_edge);
+                    }
+                }
+            }
+            // After going over all edges
+            // Were there any good swap choices?
+            if cost_reduced {
+                qubit_set.remove(&optimal_start_qubit);
+                qubit_set.remove(&optimal_end_qubit);
+                trial_layout = optimal_layout.clone();
+                opt_edges.add(optimal_start, optimal_end);
+            } else {
+                break;
+            }
+        }
+        // We have either run out of swap pairs to try or failed to improve
+        // the cost
+
+        // Compute the coupling graph distance
+        dist = compute_cost(&cdist, &trial_layout, gates, num_gates);
+        // If all gates can be applied now we're finished.
+        // Otherwise we need to consider a deeper swap circuit
+        if dist as usize == num_gates {
+            break;
+        }
+        // increment the depth
+        depth_step += 1;
+    }
+    // Either we have succeeded at some depth d < d_max or failed
+    dist = compute_cost(&cdist, &trial_layout, gates, num_gates);
+    if dist as usize == num_gates && depth_step == 1 {
+        let mut best_possible = locked_best_possible.write().unwrap();
+        **best_possible = Some((dist, opt_edges, trial_layout, depth_step));
+        return None;
+    }
+    Some((dist, opt_edges, trial_layout, depth_step))
+}
+
+#[pyfunction]
+#[pyo3(text_signature = "(graph, weight_fn, /)")]
+pub fn swap_trials<'p>(
```

Comment:
What's the generic type over here?

Reply:
This is actually a lifetime annotation (which is indicated by the ``'`` prefix), it's used to tell rust how long something is expected to live (it's kind of like scoping on steroids for references). Normally the rust compiler can elide this based on context and you don't need to explicitly set a lifetime, but in some contexts where your working with references you need to provide some more clues on how long references last so the borrow checker can reason about it. The canonical example from the rust book is:

```rust
fn longest<'a>(x: &'a str, y: &'a str) -> &'a str {
    if x.len() > y.len() {
        x
    } else {
        y
    }
}
```
(from https://doc.rust-lang.org/book/ch10-03-lifetime-syntax.html#lifetime-annotation-syntax )
where the lifetime annotation is basically saying you can expect the reference to `x`, `y`, and the returned referenced all will be valid for at least the same lifetime `'a`.

In this context, I can check it might not be explicitly necessary. Normally in retworkx I had to do this because the functions which are working with readonly numpy views also take a `Python` parameter (which is basically a wrapper around a GIL handle) and I need to explicitly specify the lifetime to tell the compiler that lifetime of the internal reference for the numpy array view will be at least as long as the lifetime for the gil handle. But in this context since we're not using the gil at all it may not actually be necessary and the compiler can figure it out on it's own. I'll give dropping the lifetime here a try and see what happens.

Reply:
Ohhh, my bad.  I know the principles of lifetimes, I just mixed things up from my F# days, where generic parameters are introduced with the apostrophe - I'm used to writing things like
```f#
let fold (f: 'State -> 'T -> 'State) (state: 'State) xs =
    match xs
    | [] -> state
    | [x] -> f state x
    | head :: tail -> fold f (f state head) tail
```

Reply:
I was right, we didn't actually need it here and I removed the lifetime annotation in: https://github.com/Qiskit/qiskit-terra/pull/7658/commits/780f327fa09db5a4e701f7a4c1b24120c52ebdc1